Add a minimum limit to Auto refresh intervals

Thib17 opened the pull request:
Using grafana with a lot of users and dashboards may raises some issues like
making too many requests against data sources. As far as I see, this is often
due to users who defined the auto-refresh interval too low (e.g. auto-refresh
set as 5s even if the retention policy is set at 1m).
This change adds a config input to set a minimum limit to auto-refresh intervals
available.

torkelo commented:
looks interesting! 

But I do not think this should be a global server setting but a data source setting (and a dashboard setting). 

And might be good to include both a auto refresh limit and a time range limit.

Thib17 commented:
Just to make sure, do you mean having this setting for each data source ? A dashboard can have several data sources. So a way to have a data source setting could be setting the limit for each data source and then use the maximum of the limits from data sources used in the dashboard. Is it what you had in mind ?
And about dashboard setting, do you mean allowing users to override it in their dashboards ?

torkelo commented:
Yes, using the maximum of dashboard data sources sounds like a good idea. I think local dashboard options also make sense (especially for time range limits), but the override should not allow a dashboard editor to go below the data source auto refresh limit or the above the time range limit

Commit:
Add a minimum interval per datasource for auto-refresh

Using grafana with a lot of users and dashboards may raises some issues like
making too many requests against data sources. As far as I see, this is often
due to users who defined the auto-refresh interval too low (e.g. auto-refresh
set as 5s even if the retention policy is set at 1m).
This change adds a config input per datasource to set a minimum interval for
auto-refresh. Then each dashboard use the strongest limit from all used
datasources.

Thib17 commented:
The limit is now using data source settings (not global server setting anymore). @torkelo, does it fit what you suggested ?

torkelo commented on pkg/api/datasources.go:
```diff
@@ -22,18 +22,19 @@ func GetDataSources(c *middleware.Context) {
-			BasicAuth: ds.BasicAuth,
-			IsDefault: ds.IsDefault,
-			JsonData:  ds.JsonData,
+			Id:                     ds.Id,
```
wonder why there is much indentation changes, did you run this through gofmt?

Thib17 replied:
That was due to the new column.

torkelo commented on pkg/api/datasources.go:
```diff
@@ -191,22 +192,23 @@ func GetDataSourceIdByName(c *middleware.Context) Response {
+		IsDefault:              ds.IsDefault,
+		JsonData:               ds.JsonData,
+		SecureJsonFields:       map[string]bool{},
+		MinAutoRefreshInterval: ds.MinAutoRefreshInterval,
```
do we really need a new column for this? Maybe we can use a property on the jsonData

Thib17 replied:
Yes, you're right ! (and that's a lot simplier)

Thib17 commented:
As you suggested, it uses a property on the jsonData. It's much simpler now, thx!

Thib17 commented:
Any news about this PR @torkelo ?

torkelo commented:
Sorry. need to find time to review this

Thib17 commented:
@torkelo, did you have time to look at it ?

torkelo commented:
yes, it needs more work. not happy with the amount of code in the timepicker. Think this needs to be coupled with time range limits as well. So need some thought into design.

iksaif commented:
@torkelo I guess most of the datasource code could move a a dashboard specific js file. Any other suggestion so we can make progress on this PR ? :)

torkelo commented:
This solution feels a bit clunky as it complicates the time picker component with responsibilities that should be outside it's scope. 

For now I solving this by modifying the time picker refresh options in dashboard settings is ok. 
![image](https://cloud.githubusercontent.com/assets/10999/26144863/cdba1306-3aea-11e7-9a93-48e92678fdad.png)

iksaif commented:
@torkelo  as said earlier in the thread, in order to protect his service an administrator needs to be able to set a minimum for the auto-refresh setting, else Grafana becomes a DDoS client for your TSDB pretty quickly. Doing that per-dashboard isn't really practical.

byblakout commented:
Is this currently only available per-dashboard? We have 200+ dashboards available to staff across various teams. Globally setting a minimum auto-refresh value would be extremely valuable (so far Prometheus and Graphite have been killed by dashboards).

craigwatkinson commented:
In addition to the comments above, this solution doesn't actually enforce a minimum. If a user sets a lower value in the URL, the requested refresh is allowed
